File: src/tools/write.zig
const std = @import("std");
const core = @import("phoenix_core");

const max_bytes: usize = 512 * 1024;

fn errResult(allocator: std.mem.Allocator, comptime fmt: []const u8, args: anytype) anyerror!core.tool.ToolResult {
    const msg = try std.fmt.allocPrint(allocator, fmt, args);
    return .{ .output = msg, .is_error = true };
}

fn invoke(self: *const core.Tool, io: std.Io, args: []const u8, allocator: std.mem.Allocator) anyerror!core.tool.ToolResult {
    _ = self;

    var parsed = std.json.parseFromSlice(std.json.Value, allocator, args, .{}) catch {
        return errResult(allocator, "write: invalid JSON args: {s}", .{args});
    };
    defer parsed.deinit();
    if (parsed.value != .object) return errResult(allocator, "write: args must be a JSON object", .{});
    const obj = parsed.value.object;

    const path_v = obj.get("path") orelse return errResult(allocator, "write: missing required field 'path'", .{});
    if (path_v != .string) return errResult(allocator, "write: 'path' must be a string", .{});
    const path = path_v.string;

    const content_v = obj.get("content") orelse return errResult(allocator, "write: missing required field 'content'", .{});
    if (content_v != .string) return errResult(allocator, "write: 'content' must be a string", .{});
    const content = content_v.string;

    if (std.fs.path.dirname(path)) |dir| {
        if (dir.len > 0) {
            std.Io.Dir.cwd().createDirPath(io, dir) catch |err| {
                return errResult(allocator, "write: cannot create parent directory {s}: {s}", .{ dir, @errorName(err) });
            };
        }
    }

    std.Io.Dir.cwd().writeFile(io, .{ .sub_path = path, .data = content }) catch |err| {
        return errResult(allocator, "write: cannot write {s}: {s}", .{ path, @errorName(err) });
    };

    const msg = try std.fmt.allocPrint(allocator, "wrote {d} bytes to {s}", .{ content.len, path });
    return .{ .output = msg };
}

pub const write_tool: core.Tool = .{
    .name = "write",
    .description = "Write content to a file. Creates the file if it doesn't exist, overwrites if it does. Automatically creates parent directories.",
    .schema =
    \\{"type":"object","properties":{"path":{"type":"string","description":"Path to the file to write (relative or absolute)"},"content":{"type":"string","description":"Content to write to the file"}},"required":["path","content"]}
    ,
    .max_output_bytes = max_bytes,
    .invokeFn = invoke,
};

// ---- Tests ----

test "write creates file and parent dir" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const tmp_path = "phoenix_test_write_dir/nested/dir/out.txt";
    defer {
        std.Io.Dir.cwd().deleteTree(io, "phoenix_test_write_dir") catch {};
    }

    const args =
        \\{"path":"phoenix_test_write_dir/nested/dir/out.txt","content":"hello world"}
    ;
    const r = try write_tool.invoke(io, args, a);
    defer a.free(r.output);
    try std.testing.expect(!r.is_error);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "wrote 11 bytes") != null);

    const back = try std.Io.Dir.cwd().readFileAlloc(io, tmp_path, a, .limited(128));
    defer a.free(back);
    try std.testing.expectEqualStrings("hello world", back);
}

test "write overwrites existing file" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const tmp_path = "phoenix_test_write_overwrite.txt";
    try std.Io.Dir.cwd().writeFile(io, .{ .sub_path = tmp_path, .data = "OLD CONTENT THAT IS LONGER" });
    defer std.Io.Dir.cwd().deleteFile(io, tmp_path) catch {};

    const args =
        \\{"path":"phoenix_test_write_overwrite.txt","content":"new"}
    ;
    const r = try write_tool.invoke(io, args, a);
    defer a.free(r.output);

    const back = try std.Io.Dir.cwd().readFileAlloc(io, tmp_path, a, .limited(128));
    defer a.free(back);
    try std.testing.expectEqualStrings("new", back);
}

test "write missing content is error" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const args =
        \\{"path":"/tmp/x"}
    ;
    const r = try write_tool.invoke(io, args, a);
    defer a.free(r.output);
    try std.testing.expect(r.is_error);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "missing required field 'content'") != null);
}

File: src/tools/write.rs
use async_trait::async_trait;
use serde_json::{Value, json};

use super::traits::{Tool, ToolError, ToolResult, ToolSchema};

pub struct WriteTool;

#[async_trait]
impl Tool for WriteTool {
    fn schema(&self) -> ToolSchema {
        ToolSchema {
            name: "write",
            description: "Write content to a file. Creates the file if it doesn't exist, \
                          overwrites if it does. Automatically creates parent directories.",
            parameters: json!({
                "type": "object",
                "properties": {
                    "file_path": {
                        "type": "string",
                        "description": "Absolute path to the file to write"
                    },
                    "content": {
                        "type": "string",
                        "description": "Content to write to the file"
                    }
                },
                "required": ["file_path", "content"]
            }),
        }
    }

    async fn invoke(&self, args: Value) -> Result<ToolResult, ToolError> {
        let file_path = args
            .get("file_path")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ToolError::InvalidArgs("missing required field 'file_path'".into()))?;

        let content = args
            .get("content")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ToolError::InvalidArgs("missing required field 'content'".into()))?;

        // Create parent directories if needed.
        let path = std::path::Path::new(file_path);
        if let Some(parent) = path.parent()
            && !parent.as_os_str().is_empty()
        {
            tokio::fs::create_dir_all(parent).await.map_err(|e| {
                ToolError::ExecutionFailed(format!(
                    "cannot create parent directory {}: {e}",
                    parent.display()
                ))
            })?;
        }

        tokio::fs::write(file_path, content)
            .await
            .map_err(|e| ToolError::ExecutionFailed(format!("cannot write {file_path}: {e}")))?;

        let preview_lines: Vec<&str> = content.lines().take(30).collect();
        let total_lines = content.lines().count();
        let mut output = format!(
            "wrote {file_path} ({} bytes, {total_lines} lines)\n",
            content.len()
        );
        for line in &preview_lines {
            output.push_str(&format!("+ {line}\n"));
        }
        if total_lines > 30 {
            output.push_str(&format!("... ({} more lines)\n", total_lines - 30));
        }
        Ok(ToolResult::success(output))
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    #[tokio::test]
    async fn write_creates_file() {
        let dir = TempDir::new().unwrap();
        let path = dir.path().join("out.txt");
        let path_str = path.to_str().unwrap();

        let result = WriteTool
            .invoke(json!({"file_path": path_str, "content": "hello world"}))
            .await
            .unwrap();

        assert!(!result.is_error);
        assert!(result.output.contains("wrote"));
        assert!(result.output.contains("11 bytes"));
        assert_eq!(std::fs::read_to_string(&path).unwrap(), "hello world");
    }

    #[tokio::test]
    async fn write_creates_nested_parent_dirs() {
        let dir = TempDir::new().unwrap();
        let path = dir.path().join("a").join("b").join("c").join("file.txt");
        let path_str = path.to_str().unwrap();

        let result = WriteTool
            .invoke(json!({"file_path": path_str, "content": "nested"}))
            .await
            .unwrap();

        assert!(!result.is_error);
        assert_eq!(std::fs::read_to_string(&path).unwrap(), "nested");
    }

    #[tokio::test]
    async fn write_overwrites_existing() {
        let dir = TempDir::new().unwrap();
        let path = dir.path().join("overwrite.txt");
        std::fs::write(&path, "OLD CONTENT THAT IS LONGER").unwrap();
        let path_str = path.to_str().unwrap();

        let result = WriteTool
            .invoke(json!({"file_path": path_str, "content": "new"}))
            .await
            .unwrap();

        assert!(!result.is_error);
        assert_eq!(std::fs::read_to_string(&path).unwrap(), "new");
    }

    #[tokio::test]
    async fn write_missing_content() {
        let result = WriteTool.invoke(json!({"file_path": "/tmp/x"})).await;
        match result {
            Err(ToolError::InvalidArgs(msg)) => {
                assert!(msg.contains("content"), "msg was: {msg}");
            }
            other => panic!("expected InvalidArgs, got: {other:?}"),
        }
    }

    #[tokio::test]
    async fn write_missing_path() {
        let result = WriteTool.invoke(json!({"content": "data"})).await;
        match result {
            Err(ToolError::InvalidArgs(msg)) => {
                assert!(msg.contains("file_path"), "msg was: {msg}");
            }
            other => panic!("expected InvalidArgs, got: {other:?}"),
        }
    }
}
